cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClientService.java

```diff
@@ -342,13 +366,18 @@ public void close() throws IOException {
       super.close();
       if (!lastProgramStateMessages.isEmpty()) {
         try {
+          LOG.debug("Sending {} program completion messages to {}",
+                    lastProgramStateMessages.size(), topicId);
           Retries.runWithRetries(() -> super.processMessages(lastProgramStateMessages.iterator()),
                                  getRetryStrategy(), t -> t instanceof IOException || t instanceof RetryableException);
         } catch (BadRequestException e) {
           // This shouldn't happen. If it does, that means the server thinks this program is no longer running.
           // The best we can do is to log here, even the log won't be collected by CDAP, but it will be retained
           // on the cluster.
           LOG.warn("Bad request when program state messages to runtime server: {}", lastProgramStateMessages, e);
+        } catch (GoneException e) {
+          LOG.warn("Failed to publish program state messages to {} since the run already marked completed",
```

Comment:
nit: "run already -> run is already " if changed for the above LOG.warn

Reply:
done